server/src/test/java/org/elasticsearch/cluster/routing/allocation/SameShardRoutingTests.java

```diff
@@ -136,6 +139,109 @@ public void testSameHost() {
+                            "node1",
+                            "node1",
+                            host1,
+                            host1,
+                            buildNewFakeTransportAddress(),
+                            emptyMap(),
+                            MASTER_DATA_ROLES,
+                            Version.CURRENT
+                        )
+                    )
+                    .add(
+                        new DiscoveryNode(
+                            "node2",
+                            "node2",
+                            "node2",
+                            host2,
+                            host2,
+                            buildNewFakeTransportAddress(),
+                            emptyMap(),
+                            MASTER_DATA_ROLES,
+                            Version.CURRENT
+                        )
+                    )
+            )
+            .build();
+        clusterState = applyStartedShardsUntilNoChange(clusterState, strategy);
+
+        List<ShardRouting> unassignedShards = shardsWithState(clusterState.getRoutingNodes(), UNASSIGNED);
+        if (host1.equals(host2) == false) {
+            logger.info("Two nodes on the different host");
+            assertThat(0, equalTo(unassignedShards.size()));
+        } else {
+            final ShardRouting unassignedShard = unassignedShards.get(0);
+
+            final SameShardAllocationDecider decider = new SameShardAllocationDecider(
+                sameHostSetting,
+                new ClusterSettings(sameHostSetting, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS)
+            );
+
+            final RoutingNode emptyNode = StreamSupport.stream(clusterState.getRoutingNodes().spliterator(), false)
+                .filter(node -> node.getByShardId(unassignedShard.shardId()) == null)
+                .findFirst()
+                .orElseThrow(AssertionError::new);
+
+            final RoutingAllocation routingAllocation = new RoutingAllocation(
+                new AllocationDeciders(singletonList(decider)),
+                clusterState.getRoutingNodes(),
+                clusterState,
+                null,
+                null,
+                0L
+            );
+            routingAllocation.debugDecision(true);
+
+            final Decision decision = decider.canAllocate(unassignedShard, emptyNode, routingAllocation);
+            logger.info("Two nodes on the same host");
+            assertThat(decision.type(), equalTo(Decision.Type.NO));
+            assertThat(
+                decision.getExplanation(),
+                equalTo(
+                    String.format(
+                        "a copy of this shard is already allocated to host %s [%s], on node [%s], and [%s] is [true] which "
+                            + "forbids more than one node on this host from holding a copy of this shard",
+                        "address",
+                        "test1",
+                        emptyNode.nodeId(),
```

Comment:
This seems wrong, we're saying `a copy of this shard is already allocated to ... node [emptyNode]` but this node is empty.

(this PR does not introduce this bug but it does crystallize it in this incorrect assertion)